rfcs/0024-incremental-ledger-catchup.md

```diff
@@ -0,0 +1,129 @@
+       - requesting a merkle path/list from peer's oldest transition to the requested transition together with that oldest transition.
+   
+   * The merkle path/list contains a list of *state_body_hash*es as its
+*proof_elem*s. Upon received the merkle_path/list, we could verify that the merkle path by first trying
+to find the oldest transition in our frontier or root_history and then call
+*Merkle_list.verify* on that merkle path. This would guarantee that the
+peer didn't send us a list of garbage and it also guarantees that the
+order is correct. And we could then reconstruct a list of *state_hash*es
+from the list of *state_body_hash*es.[2] Using this list of *state_hash*es we
+can find the missing transitions by repeated searching *state_hash*es until
+we find one (the one we find is the closest ancestor in our transition
+frontier).
+   
+2) Depending on the size of the list of the missing transitions, spawn one or
+more parallel requests to get the missing transitions from peers. We could
+verify the list of transitions that send by peer against the list of
+state_hashes.
+
+For trust system, I would describe different actions for the 2 phases described
+above:
+1) * If a peer isn't able to handle the request, we shouldn't decrease their
+     trust score.
+   * If the peer send us the a merkle path that doesn't pass verification we
+     should flag the peer as **Violated_protocol**.
+   
+2) * If a peer isn't able to handle the request, we shouldn't decrease their
+     trust score.
+   * If a peer returns a list of transitions whose state_hashes are different
+     from what we have, we should flag the peer as **Violate_protocol**.
+   * If transitions returned by the peer don't pass verification or their
+     proofs are invalid, we should decrease the peer's trust score accordingly.
+
+## Drawbacks
+
+The new design has the overhead of first downloading a merkle path. But this
+overhead is negligible comparing to the overhead of downloading unnecessary
+transitions.
+
+## Unsolved questions
+
+It's not very clear to me what's a reasonable size of the list of transition
+that we should download in 1 request.
+
+## Rationale
+
+We currently have an almost stable testnet which is quite small. We are
+moving toward the direction of solving the known performance issues that would
+hinder us from scaling up to a more realistic setting.
+
+The rationale behind this proposed design is that downloading a list of
+*state_body_hash*es is much more cheaper than downloading a list of **transition*s.
+And we could use the list of *state_body_hash*es to figure out the missing *transition*s,
+therefore avoiding downloading unnecessary things.
+
+The proposed design would do much better than the current in average case. Let's see some examples first:
+
+In all of the following examples, let's assume **k** = 3000, **size of transition** = 1 kB, **size of state_body_hash** = 5 bytes, 
```

Comment:
I think `State_body_hash.t` is closer to 95 bytes (when built with `dev`). You can ask for its length in triples, multiply by 8 / 3 to get the number of bits.

Reply:
I created a `State_hash.t` value, the number of reachable words is 9. So on a 64-bit = 8 bytes machine, that's 72 bytes.